### PR TITLE
gh-112301: Add macOS warning tracking tooling

### PR DESCRIPTION
**Pull request by nohlson**
Implement warning tracking for macOS GitHub Actions builds.

This tooling builds on https://github.com/python/cpython/pull/121730

`check_warnings.py` now takes an argument to parse compiler output of two types:
- json from gcc using `-fdiagnostic-format=json`
- generic clang compiler warning output

The macOS build and test job configuration is modified to write compiler output to a file and run the `check_warnings.py` tool.

* Issue: gh-112301

### CODE REVIEW

File: Tools/build/check_warnings.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+Parses compiler output with -fdiagnostics-format=json and checks that warnings exist
+only in files that are expected to have warnings.
+"""
+import argparse
+import json
+import re
+import sys
+from pathlib import Path
+
+def extract_warnings_from_compiler_output_clang(compiler_output: str) -> list[dict]:
```
Comment (nohlson) [inline Tools/build/check_warnings.py]: This PR depends on https://github.com/python/cpython/pull/121730 merging

File: Lib/test/test_type_annotations.py [hunk 2]
```diff
@@ -1,12 +1,9 @@
+import annotationlib
```
Comment (JelleZijlstra) [inline Lib/test/test_type_annotations.py]: This PR seems to include a lot of unrelated changes; please make sure to limit the PR to only your changes.

Comment (nohlson) [reply Lib/test/test_type_annotations.py]: I somehow rebased a bunch of commits not from main. Should be good now.

**Comment by hugovk** [comment]
This has some conflicts now https://github.com/python/cpython/pull/121730 has been merged.

**Comment by nohlson** [comment]
> This has some conflicts now #121730 has been merged.

I have updated the script to included changes made in #121730

File: Tools/build/check_warnings.py [hunk 3]
```diff
@@ -114,29 +142,40 @@ def get_unexpected_improvements(
...
         "--compiler-output-file-path",
         type=str,
         required=True,
         help="Path to the compiler output file",
     )
     parser.add_argument(
+        "-i",
         "--warning-ignore-file-path",
         type=str,
         required=True,
```
Comment (nohlson) [inline Tools/build/check_warnings.py]: Make warning ignore file optional. Without warning ignore file script should assume that there are no warnings to ignore.

Comment (nohlson) [reply Tools/build/check_warnings.py]: Made this argument optional. The log will now warn that there was no log file specified and that the script will continue ignoring no warnings.

File: Tools/build/check_warnings.py [hunk 4]
```diff
@@ -9,17 +9,42 @@
...
+    Extracts warnings from the compiler output when using clang
+    """
+    # Regex to find warnings in the compiler output
+    clang_warning_regex = re.compile(
+        r"(?P<file>.*):(?P<line>\d+):(?P<column>\d+): warning: (?P<message>.*)"
+    )
+    compiler_warnings = []
+    for line in compiler_output.splitlines():
+        match = clang_warning_regex.match(line)
+        if match:
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
        if match := clang_warning_regex.match(line):
```

File: Tools/build/check_warnings.py [hunk 5]
```diff
@@ -28,13 +53,27 @@ def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
...
+            for warning in warning_list:
+                locations = warning["locations"]
+                for location in locations:
+                    for key in ["caret", "start", "end"]:
+                        if key in location:
+                            compiler_warnings.append(
+                                {
+                                    "file": location[key]["file"].lstrip(
+                                        "./"
+                                    ),  # Remove leading current directory if present
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
                                    # Remove leading current directory if present
                                    "file": location[key]["file"].lstrip("./"),
```

File: Tools/build/check_warnings.py [hunk 6]
```diff
@@ -9,17 +9,42 @@
...
+    Extracts warnings from the compiler output when using -fdiagnostics-format=json
 
-    # Regex to find json arrays at the top level of the file
-    # in the compiler output
+    Compiler output as a whole is not a valid json document, but includes many json
+    objects and may include other output that is not json.
+    """
+    # Regex to find json arrays at the top level of the file in the compiler output
     json_arrays = re.findall(
         r"\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*\])*\])*\]", compiler_output
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: PyCharm says these are redundant character escapes '\]', I've not tested it without though:

```suggestion
        r"\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*])*])*]", compiler_output
```

File: Tools/build/check_warnings.py [hunk 7]
```diff
@@ -43,35 +82,25 @@ def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
 
 def get_warnings_by_file(warnings: list[dict]) -> dict[str, list[dict]]:
     """
-    Returns a dictionary where the key is the file and the data is the warnings
-    in that file
+    Returns a dictionary where the key is the file and the data is the warnings in that file
     """
     warnings_by_file = {}
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: Perhaps replace this with a defaultdict?

Up top:

```python
from collections import defaultdict
```

```suggestion
    warnings_by_file = defaultdict(list)
```

File: Tools/build/check_warnings.py [hunk 8]
```diff
@@ -43,35 +82,25 @@ def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
...
-                    file = file.lstrip(
-                        "./"
-                    )  # Remove leading current directory if present
-                    if file not in warnings_by_file:
-                        warnings_by_file[file] = []
-                    warnings_by_file[file].append(warning)
+        file = warning["file"]
+        if file not in warnings_by_file:
+            warnings_by_file[file] = []
+        warnings_by_file[file].append(warning)
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
        file = warning["file"]
        warnings_by_file[file].append(warning)
```
Or
```suggestion
        warnings_by_file[warning["file"]].append(warning)
```

File: Tools/build/check_warnings.py [hunk 9]
```diff
@@ -43,35 +82,25 @@ def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
...
-    warnings: list[dict],
     files_with_expected_warnings: set[str],
     files_with_warnings: set[str],
 ) -> int:
     """
-    Returns failure status if warnings discovered in list of warnings
-    are associated with a file that is not found in the list of files
-    with expected warnings
+    Returns failure status if warnings discovered in list of warnings are associated with a file
+    that is not found in the list of files with expected warnings
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: Remember to keep these <= 79 chars.

File: Tools/build/check_warnings.py [hunk 10]
```diff
@@ -88,13 +117,12 @@ def get_unexpected_warnings(
...
 def get_unexpected_improvements(
-    warnings: list[dict],
     files_with_expected_warnings: set[str],
     files_with_warnings: set[str],
 ) -> int:
     """
-    Returns failure status if there are no warnings in the list of warnings for
-    a file that is in the list of files with expected warnings
+    Returns failure status if there are no warnings in the list of warnings for a file
+    that is in the list of files with expected warnings
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: <= 79 chars here too.

File: Tools/build/check_warnings.py [hunk 11]
```diff
@@ -138,8 +165,15 @@ def main(argv: list[str] | None = None) -> int:
         "--fail-on-improvement",
         action="store_true",
         default=False,
-        help="Flag to fail if files that were expected "
-        "to have warnings have no warnings",
+        help="Flag to fail if files that were expected to have warnings have no warnings",
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: <= 79

File: Tools/build/check_warnings.py [hunk 12]
```diff
@@ -149,44 +183,53 @@ def main(argv: list[str] | None = None) -> int:
...
         return 1
 
-    # Check that the warning ignore file is a valid path
-    if not Path(args.warning_ignore_file_path).is_file():
+    # Check that a warning ignore file was specified and if so is a valid path
+    if not args.warning_ignore_file_path:
         print(
-            "Warning ignore file does not exist: "
-            f"{args.warning_ignore_file_path}"
+            "Warning ignore file not specified. Continuing without it (no warnings ignored)."
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: <= 79

File: Tools/build/check_warnings.py [hunk 13]
```diff
@@ -149,44 +183,53 @@ def main(argv: list[str] | None = None) -> int:
     # Check that the compiler output file is a valid path
     if not Path(args.compiler_output_file_path).is_file():
         print(
-            "Compiler output file does not exist: "
-            f"{args.compiler_output_file_path}"
+            f"Compiler output file does not exist: {args.compiler_output_file_path}"
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: <= 79

File: Tools/build/check_warnings.py [hunk 14]
```diff
@@ -149,44 +183,53 @@ def main(argv: list[str] | None = None) -> int:
...
-            "Warning ignore file does not exist: "
-            f"{args.warning_ignore_file_path}"
+            "Warning ignore file not specified. Continuing without it (no warnings ignored)."
         )
-        return 1
+        files_with_expected_warnings = set()
+    else:
+        if not Path(args.warning_ignore_file_path).is_file():
+            print(
+                f"Warning ignore file does not exist: {args.warning_ignore_file_path}"
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: <= 79

File: Tools/build/check_warnings.py [hunk 15]
```diff
@@ -9,17 +9,42 @@
...
+) -> list[dict]:
     """
+    Extracts warnings from the compiler output when using -fdiagnostics-format=json
 
-    # Regex to find json arrays at the top level of the file
-    # in the compiler output
+    Compiler output as a whole is not a valid json document, but includes many json
+    objects and may include other output that is not json.
+    """
+    # Regex to find json arrays at the top level of the file in the compiler output
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: <= 79

File: Tools/build/check_warnings.py [hunk 16]
```diff
@@ -149,44 +183,53 @@ def main(argv: list[str] | None = None) -> int:
...
+    elif args.compiler_output_type == "clang":
+        warnings = extract_warnings_from_compiler_output_clang(
+            compiler_output_file_contents
+        )
+
     files_with_warnings = get_warnings_by_file(warnings)
 
     status = get_unexpected_warnings(
-        warnings, files_with_expected_warnings, files_with_warnings
+        files_with_expected_warnings, files_with_warnings
```
Comment (hugovk) [inline Tools/build/check_warnings.py:238]: My IDE says "Expected type 'set[str]', got 'dict[str, list[dict]]' instead" (and for line 232 as well).

Do the functions' hints need updating to `dict`?

**Review by hugovk — COMMENTED** [review]
Looks good, just some nits :)

File: Tools/build/check_warnings.py (hunk 4 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Moved assignment to in if condition

File: Tools/build/check_warnings.py (hunk 5 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Consolidated into single line

File: Tools/build/check_warnings.py (hunk 6 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: I was able to find an even simpler regex that captures the warnings in my sample outputs. Let me know if you have any other improvements.

File: Tools/build/check_warnings.py (hunk 7 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Good idea. Switched to defaultdict to avoid checking if keys exist

File: Tools/build/check_warnings.py (hunk 8 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Implemented the latter

File: Tools/build/check_warnings.py (hunk 12 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Reformatted for character length

File: Tools/build/check_warnings.py (hunk 13 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Reformatted for character length

File: Tools/build/check_warnings.py (hunk 14 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Reformatted for character length

File: Tools/build/check_warnings.py (hunk 15 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Reformatted for character length

File: Tools/build/check_warnings.py (hunk 16 above)
Comment (nohlson) [reply Tools/build/check_warnings.py:238]: Good catch. I changed both `get_unexpected_warnings()` and `get_unexpected_improvements()` function hints to the dict

File: Tools/build/check_warnings.py [hunk 17]
```diff
@@ -2,39 +2,87 @@
...
+    Extracts warnings from the compiler output when using
+    -fdiagnostics-format=json.
 
-    # Regex to find json arrays at the top level of the file
-    # in the compiler output
+    Compiler output as a whole is not a valid json document,
+    but includes many json objects and may include other output
+    that is not json.
+    """
+    # Regex to find json arrays at the top level of the file in the compiler output
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
    # Regex to find json arrays at the top level of the file
    # in the compiler output
```

File: Tools/build/check_warnings.py [hunk 18]
```diff
@@ -2,39 +2,87 @@
...
-            )
+            warning_list = [
+                entry
+                for entry in json_objects_in_array
+                if entry.get("kind") == "warning"
+            ]
+            for warning in warning_list:
+                locations = warning["locations"]
+                for location in locations:
+                    found_location = False
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: This doesn't appear to be used:
```suggestion
```

File: Tools/build/check_warnings.py [hunk 19]
```diff
@@ -43,35 +91,22 @@ def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
 
 def get_warnings_by_file(warnings: list[dict]) -> dict[str, list[dict]]:
     """
-    Returns a dictionary where the key is the file and the data is the warnings
-    in that file
+    Returns a dictionary where the key is the file and the data is the warnings in that file
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
    Returns a dictionary where the key is the file and the data is the warnings
    in that file
```

File: Tools/build/check_warnings.py [hunk 20]
```diff
@@ -43,35 +91,22 @@ def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
...
     files_with_expected_warnings: set[str],
-    files_with_warnings: set[str],
+    files_with_warnings: dict[str, list[dict]],
 ) -> int:
     """
-    Returns failure status if warnings discovered in list of warnings
-    are associated with a file that is not found in the list of files
-    with expected warnings
+    Returns failure status if warnings discovered in list of warnings are associated with a file
+    that is not found in the list of files with expected warnings
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
    Returns failure status if warnings discovered in list of warnings
    are associated with a file that is not found in the list of files
```

File: Tools/build/check_warnings.py [hunk 21]
```diff
@@ -88,13 +123,12 @@ def get_unexpected_warnings(
...
-    warnings: list[dict],
     files_with_expected_warnings: set[str],
-    files_with_warnings: set[str],
+    files_with_warnings: dict[str, list[dict]],
 ) -> int:
     """
-    Returns failure status if there are no warnings in the list of warnings for
-    a file that is in the list of files with expected warnings
+    Returns failure status if there are no warnings in the list of warnings for a file
+    that is in the list of files with expected warnings
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
    Returns failure status if there are no warnings in the list of warnings for
    a file that is in the list of files with expected warnings
```

File: Tools/build/check_warnings.py [hunk 22]
```diff
@@ -138,8 +171,15 @@ def main(argv: list[str] | None = None) -> int:
         "--fail-on-improvement",
         action="store_true",
         default=False,
-        help="Flag to fail if files that were expected "
-        "to have warnings have no warnings",
+        help="Flag to fail if files that were expected to have warnings have no warnings",
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
        help="Flag to fail if files that were expected 
        "to have warnings have no warnings",
```

File: Tools/build/check_warnings.py (hunk 17 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Wrapped comment

File: Tools/build/check_warnings.py (hunk 18 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Removed unused variable

File: Tools/build/check_warnings.py (hunk 19 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Wrapped comment

File: Tools/build/check_warnings.py (hunk 20 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Wrapped comment

File: Tools/build/check_warnings.py (hunk 21 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Wrapped comment

File: Tools/build/check_warnings.py (hunk 22 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Wrapped string

**Comment by nohlson** [comment]
I have made the requested changes; please review again

**Review by hugovk — APPROVED** [review]
Thanks!

(The failing Hypothesis CI check is unrelated, and already fixed in `main`, we can ignore it here.)
